contracts/price_oracle/price_feed/DualCrossReferencePriceFeed.sol

```diff
@@ -0,0 +1,114 @@
+
+    /// @dev Denominator used to normalize price decimals
+    uint256 private immutable PRICE_DENOMINATOR;
+
+    /// @dev Flag indicating if we need to multiply or divide by the denominator
+    bool private immutable SHOULD_MULTIPLY;
+
+    /// @notice Constructor to initialize the price feed
+    /// @param assetX_ Asset for which the price feed is provided in USD
+    /// @param assetXAssetYOracleFeed_ Oracle feed for ASSET_X/ASSET_Y
+    /// @param assetYUsdOracleFeed_ Oracle feed for ASSET_Y/USD
+    constructor(address assetX_, address assetXAssetYOracleFeed_, address assetYUsdOracleFeed_) {
+        if (assetX_ == address(0) || assetXAssetYOracleFeed_ == address(0) || assetYUsdOracleFeed_ == address(0)) {
+            revert ZeroAddress();
+        }
+
+        ASSET_X = assetX_;
+        ASSET_X_ASSET_Y_ORACLE_FEED = assetXAssetYOracleFeed_;
+        ASSET_Y_USD_ORACLE_FEED = assetYUsdOracleFeed_;
+
+        uint256 assetXAssetYOracleFeedDecimals = AggregatorV3Interface(ASSET_X_ASSET_Y_ORACLE_FEED).decimals();
+        uint256 assetYUsdOracleFeedDecimals = AggregatorV3Interface(ASSET_Y_USD_ORACLE_FEED).decimals();
+
+        if (assetXAssetYOracleFeedDecimals < 8 || assetYUsdOracleFeedDecimals < 8) {
+            revert InvalidDecimals();
+        }
+
+        uint256 totalFeedDecimals = assetXAssetYOracleFeedDecimals + assetYUsdOracleFeedDecimals;
+        uint256 targetDecimals = _decimals();
+
+        if (totalFeedDecimals >= targetDecimals) {
+            // Need to divide by denominator
+            SHOULD_MULTIPLY = false;
+            PRICE_DENOMINATOR = 10 ** (totalFeedDecimals - targetDecimals);
+        } else {
+            // Need to multiply by denominator
+            SHOULD_MULTIPLY = true;
+            PRICE_DENOMINATOR = 10 ** (targetDecimals - totalFeedDecimals);
+        }
+    }
+
+    /// @inheritdoc IPriceFeed
+    function decimals() external pure override returns (uint8) {
+        return _decimals();
+    }
+
+    /// @inheritdoc IPriceFeed
+    function latestRoundData()
+        external
+        view
+        override
+        returns (uint80 roundId, int256 price, uint256 startedAt, uint256 time, uint80 answeredInRound)
+    {
+        (
+            uint80 assetYUsdRoundId,
+            int256 assetYPriceInUsd,
+            uint256 assetYStartedAt,
+            uint256 assetYUpdatedAt,
+            uint80 assetYAnsweredInRound
+        ) = AggregatorV3Interface(ASSET_Y_USD_ORACLE_FEED).latestRoundData();
+
+        (
+            uint80 assetXYRoundId,
+            int256 assetXPriceInAssetY,
+            uint256 assetXYStartedAt,
+            uint256 assetXYUpdatedAt,
+            uint80 assetXYAnsweredInRound
+        ) = AggregatorV3Interface(ASSET_X_ASSET_Y_ORACLE_FEED).latestRoundData();
+
+        if (assetYPriceInUsd <= 0 || assetXPriceInAssetY <= 0) revert NegativeOrZeroPrice();
+
+        uint256 rawPrice = assetYPriceInUsd.toUint256() * assetXPriceInAssetY.toUint256();
```

Comment:
Wouldn't it be easier to multiply those values and then cast to uint, or change PRICE_DENOMINATOR to int and perform calculations on them?